tmk_core/common/chibios/eeprom_wb32.c

```diff
@@ -0,0 +1,302 @@
+static uint32_t FMC_ProgramPage(uint32_t Page_Address) {
+    uint32_t ret;
+    int      state;
+
+    state = __get_PRIMASK();
+    __disable_irq();
+    PRE_OP();
+    FMC->ADDR = Page_Address;
+    ret       = FLASH_OP_EXEC(0x0C);
+    if (!state) {
+        __enable_irq();
+    }
+
+    return ret;
+}
+
+/*****************************************************************************
+ *  Erase a specified FLASH page.
+ ******************************************************************************/
+static uint32_t FMC_ErasePage(uint32_t Page_Address) {
+    uint32_t ret;
+    int      state;
+
+    state = __get_PRIMASK();
+    __disable_irq();
+    PRE_OP();
+    FMC->ADDR = Page_Address;
+    ret       = FLASH_OP_EXEC(0x08);
+    if (!state) {
+        __enable_irq();
+    }
+
+    return ret;
+}
+
+/*****************************************************************************
+ *  Clear page latches.
+ ******************************************************************************/
+static uint32_t FMC_ClearPageLatch(void) {
+    uint32_t ret;
+    int      state;
+
+    state = __get_PRIMASK();
+    __disable_irq();
+    ret = FLASH_OP_EXEC(0x04);
+    if (!state) {
+        __enable_irq();
+    }
+
+    return ret;
+}
+
+/*****************************************************************************
+ *  Erase the whole reserved Flash Space used for user Data.
+ ******************************************************************************/
+static void EEPROM_Erase(void) {
+    int page_num = 0;
+
+    eeprom_println("EEPROM_Erase");
+
+    /* delete all pages from specified start page to the last page.*/
+    do {
+        FMC_ErasePage(FEE_PAGE_BASE_ADDRESS + (page_num * FEE_PAGE_SIZE));
+        page_num++;
+    } while (page_num < FEE_PAGE_COUNT);
+}
+
+/*****************************************************************************
+ *  Read flash data.
+ ******************************************************************************/
+static void WBFLASH_Read(uint32_t ReadAddr, uint8_t *ReadDest, uint16_t ReadLen) {
+    uint16_t i;
+
+    for (i = 0; i < ReadLen; i++) {
+        ReadDest[i] = (__IO uint8_t)(*(__IO uint8_t *)ReadAddr);
+        ReadAddr++;
+    }
+}
+/*****************************************************************************
+ *  Writes once data byte to flash on specified address. If a byte is already
+ *  written, the whole page must be copied to a buffer, the byte changed and
+ *  the manipulated buffer written after PageErase.
+ *******************************************************************************/
+static uint8_t wb_flash_buffer[FEE_PAGE_SIZE] = {0};
+static int EEPROM_WriteDataPage(uint16_t Addr, uint8_t *WriteSrc, size_t Len) {
+    uint16_t     PageReamin;
+    uint16_t     PageOff;
+    uint32_t     PagePos;
+    uint32_t     WriteAddr;
+    uint32_t     WriteLen;
+
+    /* exit if desired address is above the limit (e.G. under 2048 Bytes for 4 pages).*/
```

Comment:
Also this comment seems to imply that the page size is 512 bytes.

Reply:
It's my fault, it's been modified.